hollow/src/main/java/com/netflix/hollow/api/consumer/metrics/AbstractRefreshMetricsListener.java

```diff
@@ -0,0 +1,178 @@
+    private ConsumerRefreshLoadType refreshLoadType;                // "init" indicates first successful load on consumer, "subsequent" indicates subsequent load during refresh
+
+    private long refreshLoadCount;
+
+    // Encapsulate the refresh metrics so that more metrics/properties can be added in future wihtout breaking existing clients.
+    public static class ConsumerRefreshMetrics {
+        private long refreshDurationMillis;
+        private ConsumerRefreshStatus refreshStatus;
+        private ConsumerRefreshTransitionType refreshTransitionType;
+        private ConsumerRefreshLoadType refreshLoadType;
+
+        public ConsumerRefreshMetrics(long refreshDurationMillis, ConsumerRefreshStatus refreshStatus, ConsumerRefreshTransitionType refreshTransitionType, ConsumerRefreshLoadType refreshLoadType) {
+            this.refreshDurationMillis = refreshDurationMillis;
+            this.refreshStatus = Objects.requireNonNull(refreshStatus);
+            this.refreshTransitionType = refreshTransitionType;   // Transition type can be unknown for some failed snapshot or delta transition
+            this.refreshLoadType = Objects.requireNonNull(refreshLoadType);
+        }
+
+        public long getRefreshDurationMillis() {
+            return refreshDurationMillis;
+        }
+        public ConsumerRefreshStatus getRefreshStatus() {
+            return refreshStatus;
+        }
+        public ConsumerRefreshTransitionType getRefreshTransitionType() {
+            return refreshTransitionType;
+        }
+        public ConsumerRefreshLoadType getRefreshLoadType() {
+            return refreshLoadType;
+        }
+    }
+
+    public enum ConsumerRefreshTransitionType {
+        Snapshot("snapshot"),       // refresh involved snapshot transition
+        Delta("delta");             // refresh applied a delta transition
+
+        final private String type;
+        ConsumerRefreshTransitionType(String type) {
+            this.type = type;
+        }
+    }
+
+    public enum ConsumerRefreshLoadType {
+        Init("init"),               // consumer's first-time initialized with a snapshot
+        Subsequent("subsequent");   // consumer subsequent updated with either delta or snapshot
+
+        final private String type;
+        ConsumerRefreshLoadType(String type) {
+            this.type = type;
+        }
+    }
+
+    public enum ConsumerRefreshStatus {
+        Success("success"),
+        Failure("failure");
+
+        final private String status;
+        ConsumerRefreshStatus(String status) {
+            this.status = status;
+        }
+    }
+
+
+    public AbstractRefreshMetricsListener() {
+        lastRefreshInstant = Instant.now(); // TODO: Is this an acceptable value for the first transition?
+        refreshLoadCount = 0l;
+        consecutiveSuccesses = 0l;
+        consecutiveFailures = 0l;
+        refreshStatus = ConsumerRefreshStatus.Success;  // TODO: Ok to initialize as success for first refresh?
```

Comment:
In our internal dashboard, we support showing an initial load that fails to succeed. The load type isn't necessarily the load that happened...until the moment of success it's the load/refresh that is being attempted.

So we either want to omit this value from being reported until success or failure happens, or we need to model a 3rd indeterminate state. I think I'm more in favor of the former.

Reply:
Need to chat about this one

Reply:
Success/failure will be tagged on the duration metric at refresh end, and lastRefreshTimeNanoOptional is.. well.. optional :)